jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,343 @@
+     * @return the {@link Content.Source}.
+     */
+    public static Content.Source asContentSource(Resource resource, ByteBufferPool bufferPool, int bufferSize, boolean direct)
+    {
+        if (resource.isDirectory() || !resource.exists())
+            throw new IllegalArgumentException("Resource must exist and cannot be a directory: " + resource);
+
+        // Try to find an optimized content source.
+        Path path = resource.getPath();
+        if (path != null)
+        {
+            PathContentSource pathContentSource = new PathContentSource(path, bufferPool);
+            if (bufferSize > 0)
+            {
+                pathContentSource.setBufferSize(bufferSize);
+                pathContentSource.setUseDirectByteBuffers(direct);
+            }
+            return pathContentSource;
+        }
+        if (resource instanceof MemoryResource memoryResource)
+        {
+            byte[] bytes = memoryResource.getBytes();
+            return new ByteBufferContentSource(ByteBuffer.wrap(bytes));
+        }
+
+        // Fallback to wrapping InputStream.
+        try
+        {
+            return new InputStreamContentSource(resource.newInputStream());
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    /**
+     * Gets a {@link Content.Source} with a range of the contents of a resource, if possible.
+     * Non-existent and directory resources have no content, so calling this method on such resource
+     * throws {@link IllegalArgumentException}.
+     *
+     * @param resource the resource from which to get a {@link Content.Source}.
+     * @param bufferPool the {@link ByteBufferPool} to get buffers from. null means allocate new buffers as needed.
+     * @param bufferSize the size of the buffer to be used for the copy. Any value &lt; 1 means use a default value.
+     * @param direct the directness of the buffers, this parameter is ignored if {@code bufferSize} is &lt; 1.
+     * @param first the first byte from which to read from.
+     * @param length the length of the content to read.
+     * @return the {@link Content.Source}.
+     */
+    public static Content.Source asContentSource(Resource resource, ByteBufferPool bufferPool, int bufferSize, boolean direct, long first, long length)
+    {
+        // Try using the resource's path if possible, as the nio API is async and helps to avoid buffer copies.
+        Path path = resource.getPath();
+        if (path != null)
+        {
+            RangedPathContentSource contentSource = new RangedPathContentSource(path, bufferPool, first, length);
+            if (bufferSize > 0)
+            {
+                contentSource.setBufferSize(bufferSize);
+                contentSource.setUseDirectByteBuffers(direct);
+            }
+            return contentSource;
+        }
+
+        // TODO MemoryResource could be optimized too
```

Comment:
Can you do this now, else it will be forgotten.

Reply:
Done.